File: WanderIQKit/Sources/WanderIQKit/Sync/RemoteSyncBackend.swift
import Foundation

/// One page of pulled changes plus the advanced cursor.
public struct ChangePage: Equatable, Sendable {
    public let records: [SyncRecord]
    public let cursor: Date
    public init(records: [SyncRecord], cursor: Date) {
        self.records = records; self.cursor = cursor
    }
}

/// Transport abstraction. Sub-project 3 implements this with supabase-swift
/// (PostgREST upserts + cursor query + Realtime). The engine depends only on
/// this protocol, so it is fully testable with a fake.
public protocol RemoteSyncBackend: Sendable {
    /// Upsert records (tombstones included) to the server.
    func send(_ records: [SyncRecord]) async throws
    /// Fetch records with server_updated_at strictly greater than `cursor`,
    /// and the new cursor (max server_updated_at seen, else `cursor`).
    func changes(since cursor: Date) async throws -> ChangePage
}

File: webapp/src/sync/remoteSyncBackend.ts
import { entityKey, type SyncRecord } from "./types";

export interface ChangePage {
  records: SyncRecord[];
  cursor: number;
}

/** Transport abstraction. The Supabase implementation arrives in 4b. */
export interface RemoteSyncBackend {
  send(records: SyncRecord[]): Promise<void>;
  changes(since: number): Promise<ChangePage>;
}

/** In-memory backend for tests; monotonic server clock models server_updated_at. */
export class FakeRemoteBackend implements RemoteSyncBackend {
  private stored = new Map<string, { record: SyncRecord; serverAt: number }>();
  private clock = 0;

  async send(records: SyncRecord[]): Promise<void> {
    for (const r of records) {
      this.clock += 1;
      this.stored.set(entityKey(r), { record: r, serverAt: this.clock });
    }
  }

  async changes(since: number): Promise<ChangePage> {
    const fresh = [...this.stored.values()]
      .filter((e) => e.serverAt > since)
      .sort((a, b) => a.serverAt - b.serverAt);
    const cursor = fresh.length ? fresh[fresh.length - 1].serverAt : since;
    return { records: fresh.map((e) => e.record), cursor };
  }
}
